`endpoint=127.138.100.38 is not marked for removal but owns no tokens` in `test_mutation_schema_change`

Weird failure of one of topology tests:

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1069/artifact/testlog/x86_64/dev/topology.test_mutation_schema_change.1.log
```python3
    async def test_mutation_schema_change(manager, random_tables):
        """
            Cluster A, B, C
            create table
            stop C
            change schema + do lwt write + change schema
            stop B
            start C
            do lwt write to the same key through C
        """
        server_a, server_b, server_c = await manager.running_servers()
        t = await random_tables.add_table(ncolumns=5, pks=1)
        manager.driver_close()
        # Reduce the snapshot thresholds
        await manager.mark_dirty()
        errs = [inject_error_one_shot(manager.api, s.ip_addr, 'raft_server_snapshot_reduce_threshold')
                for s in [server_a, server_b, server_c]]
        await asyncio.gather(*errs)
    
    
        logger.info("Stopping C %s", server_c)
        await manager.server_stop_gracefully(server_c.server_id)
        await manager.driver_connect()
    
>       async with inject_error(manager.api, server_b.ip_addr, 'paxos_error_before_learn'):
```

The `inject_error` call happened because the connection was dropped:
```
E               aiohttp.client_exceptions.ClientOSError: [Errno 104] Connection reset by peer
```

I looked into the logs, and it turns out that the server had crashed:
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1069/artifact/testlog/x86_64/dev/scylla-59.log
```
ERROR 2023-05-05 18:08:02,779 [shard 0] storage_service - endpoint=127.138.100.38 is not marked for removal but owns no tokens, at: /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x527f9e /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x528580 /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x528898 /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x37e983 0x21f67fa 0x21fe4fc 0x22018e3 0x20e673e 0x20ffc17 0x20ce226 0x20f301f 0xfc59fd /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c7d7f /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c8f47 /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c8369 /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x34198e /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x340b9e 0xf5ec1f 0xf97e10 0xf5dc5b /lib64/libc.so.6+0x2750f /lib64/libc.so.6+0x275c8 0xf5c7e4
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::coroutine::parallel_for_each<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_16>
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void>::then_impl_nrvo<seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void> >(seastar::noncopyable_function<seastar::future<void> ()>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> ()>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
Aborting on shard 0.
Backtrace:
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3b964c
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3e42e2
  /lib64/libc.so.6+0x3cb1f
  /lib64/libc.so.6+0x8ce5b
  /lib64/libc.so.6+0x3ca75
  /lib64/libc.so.6+0x267fb
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x37ea0b
  0x21f67fa
  0x21fe4fc
  0x22018e3
  0x20e673e
  0x20ffc17
  0x20ce226
  0x20f301f
  0xfc59fd
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c7d7f
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c8f47
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c8369
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x34198e
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x340b9e
  0xf5ec1f
  0xf97e10
  0xf5dc5b
  /lib64/libc.so.6+0x2750f
  /lib64/libc.so.6+0x275c8
  0xf5c7e4
------ Ending test dev/topology.test_mutation_schema_change.1::test_mutation_schema_change ------
```

It's a weird failure because the test performed no topology changes other than bootstrapping the initial 3 nodes before the crash.

Server logs and test log:
[dev.zip](https://github.com/scylladb/scylladb/files/11419103/dev.zip)

Test framework log:
[test.py.debug-release-dev-1069.log](https://github.com/scylladb/scylladb/files/11419118/test.py.debug-release-dev-1069.log)

--- a/configure.py
+++ b/configure.py
@@ -510,6 +510,7 @@ scylla_tests = set([
     'test/boost/s3_test',
     'test/boost/locator_topology_test',
     'test/boost/string_format_test',
+    'test/boost/tagged_integer_test',
     'test/manual/ec2_snitch_test',
     'test/manual/enormous_table_scan_test',
     'test/manual/gce_snitch_test',

--- a/gms/heart_beat_state.hh
+++ b/gms/heart_beat_state.hh
@@ -58,6 +58,7 @@ public:
     }
 
     void force_highest_possible_version_unsafe() noexcept {
+        static_assert(std::numeric_limits<version_type>::is_bounded);
         _version = std::numeric_limits<version_type>::max();
     }
 

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -2156,7 +2156,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
         }
     } else {
         if (owned_tokens.empty()) {
-            on_internal_error_noexcept(slogger, ::format("endpoint={} is not marked for removal but owns no tokens", endpoint));
+            on_internal_error(slogger, ::format("endpoint={} is not marked for removal but owns no tokens", endpoint));
         }
 
         if (!is_normal_token_owner) {

--- a/test/boost/tagged_integer_test.cc
+++ b/test/boost/tagged_integer_test.cc
@@ -1,0 +1,32 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+
+#include <seastar/testing/test_case.hh>
+#include "test/lib/scylla_test_case.hh"
+
+#include "utils/tagged_integer.hh"
+
+using namespace seastar;
+
+using test_tagged_int = utils::tagged_integer<struct test_int_tag, int>;
+
+SEASTAR_THREAD_TEST_CASE(test_tagged_integer_ops) {
+    int i = 17;
+    auto t = test_tagged_int(i);
+
+    BOOST_REQUIRE_EQUAL(t.value(), i);
+    BOOST_REQUIRE_EQUAL((++t).value(), ++i);
+    BOOST_REQUIRE_EQUAL((--t).value(), --i);
+    BOOST_REQUIRE_EQUAL((t++).value(), i++);
+    BOOST_REQUIRE_EQUAL((t--).value(), i--);
+    BOOST_REQUIRE_EQUAL((t + test_tagged_int(42)).value(), i + 42);
+    BOOST_REQUIRE_EQUAL((t - test_tagged_int(42)).value(), i - 42);
+    BOOST_REQUIRE_EQUAL((t += test_tagged_int(42)).value(), i += 42);
+    BOOST_REQUIRE_EQUAL((t -= test_tagged_int(42)).value(), i -= 42);
+}

--- a/test/boost/tagged_integer_test.cc
+++ b/test/boost/tagged_integer_test.cc
@@ -7,10 +7,14 @@
  */
 
 
-#include <seastar/testing/test_case.hh>
+#include <limits>
+
 #include "test/lib/scylla_test_case.hh"
+#include <seastar/testing/test_case.hh>
 
 #include "utils/tagged_integer.hh"
+#include "gms/generation-number.hh"
+#include "gms/version_generator.hh"
 
 using namespace seastar;
 
@@ -29,4 +33,25 @@ SEASTAR_THREAD_TEST_CASE(test_tagged_integer_ops) {
     BOOST_REQUIRE_EQUAL((t - test_tagged_int(42)).value(), i - 42);
     BOOST_REQUIRE_EQUAL((t += test_tagged_int(42)).value(), i += 42);
     BOOST_REQUIRE_EQUAL((t -= test_tagged_int(42)).value(), i -= 42);
+}
+
+SEASTAR_THREAD_TEST_CASE(test_tagged_integer_numeric_limits) {
+    BOOST_REQUIRE(std::numeric_limits<test_tagged_int>::is_specialized);
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::is_signed, std::numeric_limits<test_tagged_int::value_type>::is_signed);
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::is_integer, std::numeric_limits<test_tagged_int::value_type>::is_integer);
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::is_exact, std::numeric_limits<test_tagged_int::value_type>::is_exact);
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::is_bounded, std::numeric_limits<test_tagged_int::value_type>::is_bounded);
+
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::min().value(), std::numeric_limits<test_tagged_int::value_type>::min());
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<test_tagged_int>::max().value(), std::numeric_limits<test_tagged_int::value_type>::max());
+}
+
+SEASTAR_THREAD_TEST_CASE(test_gms_generation_type_numeric_limits) {
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<gms::generation_type>::min().value(), std::numeric_limits<gms::generation_type::value_type>::min());
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<gms::generation_type>::max().value(), std::numeric_limits<gms::generation_type::value_type>::max());
+}
+
+SEASTAR_THREAD_TEST_CASE(test_gms_version_type_numeric_limits) {
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<gms::version_type>::min().value(), std::numeric_limits<gms::version_type::value_type>::min());
+    BOOST_REQUIRE_EQUAL(std::numeric_limits<gms::version_type>::max().value(), std::numeric_limits<gms::version_type::value_type>::max());
 }

--- a/utils/tagged_integer.hh
+++ b/utils/tagged_integer.hh
@@ -69,11 +69,11 @@ public:
         return tagged_tagged_integer(_value - o._value);
     }
 
-    tagged_tagged_integer& operator+=(const tagged_tagged_integer& o) const {
+    tagged_tagged_integer& operator+=(const tagged_tagged_integer& o) {
         _value += o._value;
         return *this;
     }
-    tagged_tagged_integer& operator-=(const tagged_tagged_integer& o) const {
+    tagged_tagged_integer& operator-=(const tagged_tagged_integer& o) {
         _value -= o._value;
         return *this;
     }

--- a/utils/tagged_integer.hh
+++ b/utils/tagged_integer.hh
@@ -98,4 +98,19 @@ template <typename Final, typename Tag, std::integral ValueType>
     return s << x.value();
 }
 
+template <typename Final, typename Tag, std::integral ValueType>
+struct numeric_limits<utils::tagged_tagged_integer<Final, Tag, ValueType>> : public numeric_limits<ValueType> {
+    using tagged_tagged_integer_t = utils::tagged_tagged_integer<Final, Tag, ValueType>;
+    using value_limits = numeric_limits<ValueType>;
+    static_assert(numeric_limits<ValueType>::is_specialized && numeric_limits<ValueType>::is_bounded);
+
+    static constexpr tagged_tagged_integer_t min() {
+        return tagged_tagged_integer_t(value_limits::min());
+    }
+
+    static constexpr tagged_tagged_integer_t max() {
+        return tagged_tagged_integer_t(value_limits::max());
+    }
+};
+
 } // namespace std

